README.md
<p align="center">
<img src="https://img.shields.io/github/last-commit/destroyedlolo/GPMFMetersGenerator.svg?style=for-the-badge" /><br>
(no commit for long time doesn't mean this project is stalling but reached a stable state and I switched to others activities :relaxed:)
<hr/>

# GPMFMetersGenerator
Generate skickers from GoPro's GPMF

## Usage

```
GPMFMetersGenerator [options] file.MP4
```
(launch GPMFMetersGenerator with **-h** argument to have a list of reconized options)
```
$ ./GPMFMetersGenerator -h
GPMFMetersGenerator v4.03.00
(c) L.Faillie (destroyedlolo) 2022-23

GPMFMetersGenerator [-options] Video.mp4

Known options :
-s[2|3|b] : enable speed gfx (default 2d, 3: 3d, b: both)
-k[2|3] : enable speed tracker gfx (default 2d, 3: 3d)
-a : enable altitude gfx and draw curve from GoPro video
-A : enable altitude gfx and draw curve from GPX/story
-p : enable path gfx
-q : enable quality gfx
-t : enable Trekking statistics (HH:MM)
-T : enable Trekking statistics (HH:MM:SS)

-X : export telemetry as GPX file
-K : export telemetry as KML file

-G<file> : load a GPX file
-S<file> : load a story file
	Only a GPX or a story can be loadded, not both

-V : Don't generate video, keep PNG files
-Q : enforce quality by removing samples where GoP is > 500
-F : don't fail if the target directory exists
-v : turn verbose on
-d : turn debugging messages on
```

A directory with the same name of the mp4 file (without it's extention) will be created to
hold resulting images / Videos.

| :warning:WARNING:warning: : If you're upgrading from a previous version, please have a look on [Change file](CHANGES.md) ! |
| --- |

## Installation
### Dependancies 

- **Cairo** library is needed to generate PNG pics.
  - On source based distributions (Gentoo, ...), it sould be already installed if you're using a graphical environnement. Otherwise, install **x11-libs/cairo** package.
  - On binary based distributions (Debian, Suse ...), development version is needed. Please check for installation procedure [this page](https://www.cairographics.org/download/).
- **ffmpeg** binary vith **PNG** and **Quicktime** (.mov) support if you want GPMFMetersGenerator to generate videos.

### Compile from source

1. clone this repository
    ```
    git clone https://github.com/destroyedlolo/GPMFMetersGenerator
    ```
1.
    ```
    cd GPMFMetersGenerator
    ```
1. clone GoPro's [GPMF-parser](https://github.com/gopro/gpmf-parser) repository
    ```
    git clone https://github.com/gopro/gpmf-parser
    ```
1. build GPMFMetersGenerator
    ```
    make
    ```
1. copy **GPMFMetersGenerator** and **mkStory** in a directory in your PATH.
i.e :
    ```
    cp GPMFMetersGenerator mkStory /usr/local/bin
    ```
    
## Multiparts videos

Due to SD card's FAT limitation, GoPro camera split too longs videos in shorter chunks (on my Hero9, it seems earch chunk can't be longer than 3.7 GB so about 11 minutes in 1080p mode).

As of V2.03, **GPMFMetersGenerator** detects by itself the presence of additionnal part.

## Activity related stickers

![Screenshot](Images/Screenshot.png)

### Speed-O-Meter

![Speed](Images/Speed.png)

Can be based on GPS' 2D or 3D speed figure ... or even both (*mostly for testing purpose*)

As of **v4.05**, slope grade has been added.

### Speed tracker

![Speed tracker](Images/Strk.png)

*In this graphic, only 2D or 3D can be displayed.*

:scroll: Notez-bien :  Even when a **story** is loaded, only the speed of the current video is displayed : 
*reason is there is no speed in GPX specification.*

### Path

![Video only Path](Images/Path1.png)
![Video + GPX Path](Images/Path2.png)

Displays the north-facing path. By loading an external **GPX files**, the global path is displayed with the video highlighted based its telemetry.

![Video + story Path](Images/Path3.png)

:scroll: More, by loading a **story** (instead of a GPX file), all videos are highlighted on the track.<br>
:warning: only the current video is traced upon GoPro telemetry. Others are from GPX datas.

### Altitude

![Video only Altitude](Images/Altitude.png)
![Video + story Altitude](Images/Altitude2.png)

:scroll: by adding a **story**, the full session is displayed, and the part covered by videos highlighted.
Current video data are took from
- '**-a**' flag : from the GoPro meta data. 
- '**-A**' flag : from the data contained in the GPX tracking (from a GPX file or from a stories). Data of the GPX files can be smoothed if obtained from transcoding tables (WGS84, EGM96, ...) creating *unsightly "jumps"* when switching from GPX curves to GoPro data. With '-A' the graphs are only based on GPX data.

The floating dot and the value displayed always come from GoPro data.

### Trekking statistics

![Trekking statistics](Images/Stats.png)

Display **distance travelled** and **duration** since the beginning of the video *or* of the full hiking if included in a story. 

## Technical stickers

### <a name="GPSqual"></a>GPS signal quality

![DoP and Fixes](Images/DoP.png)

This sticker displays the GPS' [Dilution of Precision](https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation)) where values **< 500** are considered as good, and the signal Fix that can be :
- 0 : no lock
- 2 : 2D only lock
- 3 : 3D lock

## KML and GPX file generation

**GPMFMetersGenerator** can export GoPro's metrics into [KML](https://en.wikipedia.org/wiki/Keyhole_Markup_Language) or [GPX](https://en.wikipedia.org/wiki/GPS_Exchange_Format) file than can be included in a 3rd party GIS like **Google Map**, **Google Earth**, **OpenStreetMap** or such.

![Google Earth](Images/KML.png)

---

# Stories

A story is a way to group data from several sources : so even if a video covers only a part, telemetry stickers will reflect all the activity.

| :exclamation: | **Notez-bien :** Stories are for advanced users and are not mandatory to generate stickers. Especially, they are unuseful if you want to generate stickers for a single video :sunglasses:  |
|-------------|----------------------------|

To add a story while generating stickers, provide its file using **-S** option to **GPMFMetersGenerator**.

## Story from a set of videos

You may want to film an activity without being able to do it in one shot (also known as "*have to wait for friends on a long down hill*"). A story can be used to group telemetries of all parts, as if took from a single video.

**some obvious remarks :**
* If you're doing pauses during your activity, the **distance, path and altitude tracking will be accurate**, but the duration will reflect the difference between the beginning and time of the recording and not the cumulative time of effort.
* If you stop the recording but continue your activity, the **duration** will be right but not the distance. Altitude and path will contain straight shortcuts.

### mkStory

**mkStory** will build your story from videos only if there is no **-G** provided, so no GPX file.

```
$ mkStory GX01463[789].MP4 GX01464[0234567].MP4
*L* Reading 'GX014637.MP4'
*L* Reading 'GX014638.MP4'
*L* Reading 'GX014639.MP4'
*L* Reading 'GX014640.MP4'
*L* Reading 'GX014642.MP4'
*L* Reading 'GX014643.MP4'
*L* Reading 'GX014644.MP4'
*L* Reading 'GX014645.MP4'
*L* Reading 'GX014646.MP4'
*L* Reading 'GX014647.MP4'
```
As output, a file "**story**" is generated.

### Continuity issue

Adding **-v** option displays deltas between each video, helping to identify important jumps. 
```
$ ~/Projets/GPMFMetersGenerator/mkStory -v GX*.MP4
*I* Videos only mode
*L* Reading 'GX014637.MP4'
*L* Reading 'GX014638.MP4'
*I* gap : 00:01 min, estrangement : 0.54 m
*L* Reading 'GX014639.MP4'
*I* gap : 00:00 min, estrangement : 18.24 m
*L* Reading 'GX014640.MP4'
*I* gap : 00:05 min, estrangement : 20.67 m
```

In addition, **mkStory** will raise an error in case a video's timestamp is lower compared to the previous video one. 
```
*F* Video returning backward by 00:-24 minutes
```
Use **-F** option to force inclusion.

## Story from a GPX file

While doing long activities such as hiking, trekking, ski touring, mountain biking, or simply strolling, generally we are recording only small interesting parts and not the global session. In parallel, phones' sport tracker applications can provide complete trace of the activity.

**Stories**  attempt to establish the timeline between a GPX track coming from your phone and the GoPro videos metadata to get a global view of your activity in videos.

### data matching

Unfortunately, GPS position gathered by your phone and GoPro are not exactly the same  : satellites lost, precision issues or simply the fact the GoPro may start recording without having fully acquired satellites (use [GPS signal quality](#GPSqual) stickers to check your signal quality).

Even worst, the altitude is *calculated* by GPSes : Android's algorithm is by far less precise compared to GoPro one and tends to smooth the result.
This is leading to *jumps* when switching between GPX and GoPro curves. See GPMFMetersGenerator '**-a**' vs '**-A**' discuss above.

### mkStory

**mkStory** tries to match videos' metadata with phone provided GPX timeline.

```
$ ./mkStory -h

mkStory v3.03.00
(c) L.Faillie (destroyedlolo) 2022

mkStory [-options] video.mp4 ...
Known opts :
-G<file> : GPX of the hiking
-F : force video inclusion
-P : use position instead of GPX' timestamps

-v : turn verbose on
-d : turn debugging messages on
```

:warning:NOTEZ-BIEN:warning: : videos have to be provided in chronological order, otherwise the result will be void.

```
./mkStory -G ~/Images/Brute/20220212_Semnoz/nice.gpx ~/Images/Brute/20220212_Semnoz/GX01*.MP4
*O* Generating story in '/home/laurent/Images/Brute/20220212_Semnoz/nice.story'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013158.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013159.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013162.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013168.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013169.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013170.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013173.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013175.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013181.MP4'
*L* Reading '/home/laurent/Images/Brute/20220212_Semnoz/GX013183.MP4'
*I* results summaries
 ----------------------------------------------------------------------
| GPX : 2022-02-12T15:12:02+0100 -> 2022-02-12T16:52:33+0100
| Distance traveled : 7266 m
 --------------------------------------------------------------------------------------------------------
|    video     |         Beginning         |            End            |            Timestamps           | Status
|              | Index | Estrgmt  | CumDis | Index | Estrgmt  | CumDis |       GPX      |       GoPro    |
 --------------------------------------------------------------------------------------------------------
| GX013158.MP4 | 00227 |     12.6 |    464 | 00273 |      1.5 |    549 | 15:20 -> 15:22 | 15:23 -> 15:27 |  ok
| GX013159.MP4 | 00273 |      1.8 |    549 | 00328 |      3.0 |    649 | 15:22 -> 15:24 | 15:28 -> 15:31 |  ok
| GX013162.MP4 | 00360 |      5.3 |    703 | 00469 |      1.9 |    906 | 15:25 -> 15:30 | 15:34 -> 15:39 |  ok
| GX013168.MP4 | 00666 |      0.4 |   1282 | 00749 |      3.1 |   1457 | 15:37 -> 15:40 | 16:05 -> 16:08 |  ok
| GX013169.MP4 | 00887 |      0.8 |   1793 | 01085 |      1.0 |   2211 | 15:45 -> 15:53 | 16:13 -> 16:21 |  ok
| GX013170.MP4 | 01086 |      1.0 |   2211 | 01513 |      1.1 |   3762 | 15:53 -> 16:11 | 16:35 -> 16:41 |  ok
| GX013173.MP4 | 01698 |      1.8 |   4206 | 01812 |      2.0 |   4465 | 16:18 -> 16:22 | 17:11 -> 17:16 |  ok
| GX013175.MP4 | 01846 |      2.8 |   4559 | 01995 |      0.7 |   4894 | 16:24 -> 16:30 | 17:22 -> 17:29 |  ok
| GX013181.MP4 | 02244 |      3.0 |   5443 | 02422 |      2.9 |   5802 | 16:39 -> 16:46 | 17:41 -> 17:48 |  ok
| GX013183.MP4 | 02422 |      3.4 |   5802 | 02551 |      1.9 |   7240 | 16:46 -> 16:51 | 18:04 -> 18:09 |  ok
```
For each of the videos, the summary displays: 
- Guessed the corresponding **start and end GPX indexes** and **estrangements** between the GoPro and GPX localizations.
- **Timestamps** (*and you can see severe drifts of GPX ones due to a brain-damaged algorithm : the GoPro one is accurate as got from the GPS itself*).
- A "**quality status**" as following :
  - **ok** - seems to be good
  - **No GPS** - No satellite locked (*it's probably better to avoid this video in the story*)
  - **Partial GPS** - Missing satellite during the video
  - **Bad signal** - Weak GPS signal during the whole video
  - **Weak signal** - Weak GPS signal partially during the video
  - **not ending** - can't find an ending point for the video (generally due to the GPX to stopped before the video or a video not part of this hiking)
  - **Overlapping** - videos overlap (or due to a mistake in mkStory guesses, or because one of the video is not part of the hiking).

Bad example :
```
*I* results summaries
 ----------------------------------------------------------------------
| GPX : 2022-05-27T13:53:49+0200 -> 2022-05-27T14:46:59+0200
| Distance traveled : 4240 m
 --------------------------------------------------------------------------------------------------------
|    video     |         Beginning         |            End            |            Timestamps           | Status
|              | Index | Estrgmt  | CumDis | Index | Estrgmt  | CumDis |       GPX      |       GoPro    |
 --------------------------------------------------------------------------------------------------------
| GX013845.MP4 | 00091 |  12071.4 |    285 | 00092 |  12071.8 |    287 | 13:57 -> 13:57 | 13:58 -> 13:59 |  No GPS
| GX013846.MP4 | 00091 |  12071.4 |    285 | 00092 |  12071.8 |    287 | 13:57 -> 13:57 | 13:59 -> 14:00 |  No GPS
| GX013850.MP4 | 00091 |  12071.4 |    285 | 00092 |  12071.8 |    287 | 13:57 -> 13:57 | 14:05 -> 14:06 |  No GPS
| GX013851.MP4 | 00091 |  12071.4 |    285 | 00092 |  12071.8 |    287 | 13:57 -> 13:57 | 14:07 -> 14:07 |  No GPS
| GX013853.MP4 | 00091 |  12071.4 |    285 | 00225 |   2627.8 |    792 | 13:57 -> 14:03 | 14:09 -> 14:09 |  No GPS
| GX013854.MP4 | 00225 |   2627.8 |    792 | 00226 |   2631.7 |    798 | 14:03 -> 14:03 | 14:10 -> 14:10 |  No GPS
| GX013856.MP4 | 00225 |   2463.8 |    792 | 00226 |   2316.1 |    798 | 14:03 -> 14:03 | 14:11 -> 14:12 |  No GPS
| GX013861.MP4 | 00281 |      9.3 |   1036 | 00300 |     10.7 |   1107 | 14:05 -> 14:06 | 14:31 -> 14:32 |  Weak signal
| GX013865.MP4 | 00385 |      8.8 |   1458 | 00389 |     17.9 |   1472 | 14:09 -> 14:10 | 14:53 -> 14:54 |  Weak signal
| GX013870.MP4 | 00530 |      2.3 |   1890 | 00538 |      6.0 |   1906 | 14:16 -> 14:16 | 15:14 -> 15:15 |  ok
| GX013872.MP4 | 00568 |      2.1 |   2028 | 00569 |      4.9 |   2029 | 14:17 -> 14:17 | 15:23 -> 15:24 |  ok
| GX013875.MP4 | 00668 |      3.4 |   2431 | 00670 |      5.5 |   2439 | 14:22 -> 14:22 | 15:42 -> 15:42 |  ok
| GX013877.MP4 | 00736 |      2.5 |   2610 | 00790 |      4.1 |   2774 | 14:24 -> 14:27 | 15:50 -> 15:53 |  Weak signal
| GX013882.MP4 | 01120 |      4.3 |   3810 | 01128 |      5.4 |   3838 | 14:40 -> 14:41 | 16:38 -> 16:39 |  Weak signal
| GX013883.MP4 | 01128 |      2.4 |   3838 | 01192 |      4.0 |   4052 | 14:41 -> 14:43 | 16:39 -> 16:41 |  ok
| GX013885.MP4 | 01259 |      5.9 |   4236 | 01260 |     14.4 |   4238 | 14:46 -> 14:46 | 16:49 -> 16:49 |  ok
*W* At least Problem has been detected
```

### what to do in case of issues

#### Time vs position based guessing

By default, mkStory uses timestamps for matching.

In some cases, the GPX files do not contain reliable timestamps (for example, when using *autopause* on **Decathlon coach**, the timestamps are totally crap). In this case, use the '**-P**' flag to switch to "positioning" mode: mkStory will attempt to guess the insertion points of the videos based on positional matches.


CHANGES.md
# CHANGES / HISTORY of GPMFMetersGenerator

If you are updating from a previous version, pay particular attention to the changes with :loudspeaker: sign : they're highlighting an ascendant compatibility break.

## 4.xx

### 4.05

Add slope grade in speed widget

### 4.04

Smooth speed (average on SAMPLE's values)
Correct "both" speed gfx

### 4.03

Display seconds in trekking statistics

### 4.02

Handle commulative distance on videos only story

### 4.01

Correcte a bug on 1st video generation

### 4.00

Create stories from videos only.

## 3.xx

### 3.05

Some GPX editors (like https://gpx.studio/) don't keep the timezone and force 'Z' or ... nothing.
As fallback, try to read the timestamp without timezone.

### 3.04

- Add '**-A**' flag to force GPX data usage on altitude graphic

### 3.03

- mkStory can guess video inclusion based on timestamps

### 3.02

- Add **quality enforcement** flag to remove samples when DoP > 500

### 3.01

- Add **hicking figures** (Distance & Duration)
In a story, figures cover the full journey

### 3.00

- Add **mkStory**
- Add story support in **GPMFMetersGenerator**
- Altitude and path stickers improved when using a story
- Add **GPS quality** sticker

## 2.xx

### 2.03

Detect automatically if a video is multi-parts<br>
:loudspeaker: Consequently : **Only one video can be provided**

Handle AVC encoded video as well (GX and GH videos supported)

### 2.02

Finalize C++ migration : redesign storage class

### 2.01

GPX files are back.

### 2.00

Code has been totally redesigned in C++ to ease implementation of future new features.
As of v2.0, GPX import is temporary disabled.

:loudspeaker: **:warning: CAUTION :warning: : Arguments changed and are not compatible with 1.xx.** Have a look on `GPMFMetersGenerator -h` for the new syntax.

## 1.xx
Changes not tracked


mkStory.cpp
/* mkStory.cpp
 *
 * Synchronize GPX files with a collection of GoPro's video
 *
 * 16/05/2022 - Starting development
 */
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include <cassert>
#include <cstring>

#include <vector>
#include <iostream>

#include "datalib/Context.h"
#include "datalib/GPVideo.h"
#include "datalib/GPX.h"

#include "Version.h"

GPX *Gpx = NULL;	// external original GPX data

#if 0
double proximity = 1;	// Proximity threshold 
							// (bellow this distance, places are considered to 
#endif							// be the same.

	/* Extended GPVideo
	 * Add video's name
	 * Try to guess where the video is inserted into the global journey
	 */
class GPVideoExt : public GPVideo, public std::string {
public:
	struct closest {
		int idx;
		double distance;

		closest():idx(-1){}
	} beginning, end;

	GPVideoExt( char *n ) : GPVideo(n), std::string(n){}

		/* Search if the given coordinate is a better beginning or ending point */
	void Consider( int idx, GPSCoordinate &p ){
		if(this->beginning.idx == -1){	// Initialize with the 1st point
			this->beginning.idx = this->end.idx = idx;
			this->beginning.distance = this->getFirst().Estrangement(p);
			this->end.distance = this->getLast().Estrangement(p);
		} else {
			double dst = this->getFirst().Estrangement(p);
			if(dst < this->beginning.distance){	// New guessed beginning point
				this->beginning.idx = idx;
				this->beginning.distance = dst;

				this->end.idx = -1;			// invalidate end point
				this->end.distance = NAN;
			} else {
				dst = this->getLast().Estrangement(p);
				if(std::isnan(this->end.distance) || dst < this->end.distance){	// new guessed end point
					this->end.idx = idx;
					this->end.distance = dst;
				}
			}
		}
	}
};

std::vector<GPVideoExt> videos;

FILE *story = NULL;

int main(int argc, char *argv[]){
	bool force = false;	// force video inclusion
	bool gpxpos = false;	// use gpx' position instead of timestamps


			/* **
			 * Reading arguments
			 * **/

	int opt;
	while(( opt = getopt(argc, argv, ":vdhG:FP")) != -1){
		switch(opt){
		case 'G':
			if(Gpx){
				fputs("*F* Only one GPX file can be loaded at a time\n", stderr);
				exit(EXIT_FAILURE);
			} else {
				Gpx = new GPX(optarg);
				if(verbose)
					Gpx->Dump();

				std::string res = optarg;
				size_t pos = res.rfind(".gpx");
				if(pos != std::string::npos)
					res.erase(pos, 4);
				res += ".story";

				printf("*O* Generating story in '%s'\n", res.c_str());

				if(!(story = fopen(res.c_str(), "w"))){
					perror(res.c_str());
					exit(EXIT_FAILURE);		
				}
			}
			break;
		case 'd':	// debug implies verbose
			debug = true;
		case 'v':
			verbose = true;
			break;
#if 0
CAUTION : 'p' has been removed from getopt !!

		case 'p':
			proximity = atof(optarg);
			break;
#endif
		case 'F':
			force = true;
			break;
		case 'P':
			gpxpos = true;
			break;
		case '?':	// unknown option
			printf("unknown option: -%c\n", optopt);
		case 'h':
		case ':':	// no argument provided (or missing argument)
			puts(
				"\nmkStory v" VERSION "\n"
				"(c) L.Faillie (destroyedlolo) 2022-23\n"
				"\n"
				"mkStory [-options] video.mp4 ...\n"
				"Known opts :\n"
				"-G<file> : GPX of the hiking\n"
#if 0
				"-p<val> : Proximity threshold (default: 1m)\n"
#endif
				"-F : force video inclusion\n"
				"-P : use position instead of GPX' timestamps\n"
				"\n"
				"-v : turn verbose on\n"
				"-d : turn debugging messages on\n"
			);
			exit(EXIT_FAILURE);
			break;
		}
	}

	if(optind >= argc){
		puts("*F* No Video provided");
		exit(EXIT_FAILURE);		
	}

	if(Gpx){
			/* In this mode, GPX is the skeleton of the journey.
			 * mkStory tries to insert videos into the GPX path.
			 */
		if(verbose)
			puts("*I* GPX mode");

		/* **
		 * Read videos
		 * **/

		for(; optind < argc; optind++){
			GPVideoExt video(argv[optind]);	// load the first chunk

			if(verbose){
				video.Dump();
				printf("*I* Video distance vs GPX : min %.0fm, max %.0fm\n",
					video.getMin().Estrangement(Gpx->getMin()),
					video.getMax().Estrangement(Gpx->getMax())
				);
			}

			if( !Gpx->sameArea(video.getMin()) ||
				!Gpx->sameArea(video.getMax()) ){
					fprintf(stderr, "*W* '%s' seems outside the GPX trace (%.0fm and %.0fm)\n",
						video.c_str(),
						video.getMin().Estrangement(Gpx->getMin()),
						video.getMax().Estrangement(Gpx->getMax())
					);

					if(!force){
						fputs("*F* exiting (use -F to force this video inclusion)\n", stderr);
						exit(EXIT_FAILURE);
					}
			}
			videos.push_back(video);
		}


			/* **
			 * Guess where videos are placed in the hiking
			 * **/

		int idx;	// GPX's index

		if(!gpxpos){	// Use timestamps
			idx = 0;
			bool over = false;
			for(auto &v : videos){
				if(over){
					fprintf(stderr, "*F* '%s' Can't find ending (and it's not the last video)\n", v.c_str());
					exit(EXIT_FAILURE);
				}

				while( v.getFirst().diffTimeF( (*Gpx)[idx].getSampleTime()) > 0){
					if(++idx >= (int)Gpx->getSampleCount()){
						fprintf(stderr, "*F* '%s' Can't find beginning\n", v.c_str());
						exit(EXIT_FAILURE);
					}
				}
				v.beginning.idx = idx;
				v.beginning.distance = v.getFirst().Estrangement((*Gpx)[idx]);

				while( v.getLast().diffTimeF( (*Gpx)[idx].getSampleTime()) > 0){
					if(++idx >= (int)Gpx->getSampleCount())
						break;
				
				}
				if(idx >= (int)Gpx->getSampleCount()){
					over = true;
					idx--;	// The GoPro may have been stopped after the tracker
				}
				v.end.idx = idx;
				v.end.distance = v.getLast().Estrangement((*Gpx)[idx]);
			}
		} else {
			/* **
			 * Try to match based on positions ...
			 *
			 * The problem is informations may diverge between GoPro's and phone.
			 * GPS accuracies
			 * Time on the GPX tracking
			 *
			 * **/


				// As even timestamps are not guaranteed, videos are took individually
			for( idx = 0; idx < (int)Gpx->getSampleCount(); idx++ ){
				auto &gpx = (*Gpx)[idx];

				for(auto &v : videos)
					v.Consider( idx, gpx );
			}
		}

			/* **
			 * Displaying
			 * **/

		if(verbose){
			puts("*I* Results\n"
				"Index | Latitude   | Longitude  | Altitude    | Distance | Hour\n"
				"-----------------------------------------------------------------"
			);

			idx = 0;
			for(int vidx = 0; vidx < (int)videos.size(); vidx++){
				for(; idx < videos[vidx].beginning.idx; idx++){	// GPX before the video
					auto &gpx = (*Gpx)[idx];
					printf("%05d | %10f | %10f | %10f | %8.3f | %s\n",
						idx,
						gpx.getLatitude(), gpx.getLongitude(),
						gpx.getAltitude(), gpx.getCumulativeDistance() / 1000,
						gpx.strLocalHour().c_str()
					);
				}

				puts(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
				printf("%s\n", basename(videos[vidx].c_str()));
				printf("video's starting timestamps : %s\n", videos[vidx].getFirst().strLocalTime().c_str() );
				puts(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");

				for(; idx < videos[vidx].end.idx; idx++){	// GPX before the video
					auto &gpx = (*Gpx)[idx];
					printf("%05d | %10f | %10f | %10f | %8.3f | %s\n",
						idx,
						gpx.getLatitude(), gpx.getLongitude(),
						gpx.getAltitude(), gpx.getCumulativeDistance() / 1000,
						gpx.strLocalHour().c_str()
					);
				}
				printf("<<<<<<<<<<<<<<<<<<<<<<< %s\n", basename(videos[vidx].c_str()));
			}
	
			for(; idx < (int)Gpx->getSampleCount(); idx++){	// Remaining
				auto &gpx = (*Gpx)[idx];
				printf("%05d | %10f | %10f | %10f | %8.3f | %s\n",
					idx,
					gpx.getLatitude(), gpx.getLongitude(),
					gpx.getAltitude(), gpx.getCumulativeDistance() / 1000,
					gpx.strLocalHour().c_str()
				);
			}

			puts("");
		}

		puts("*I* results summaries");

		int prev = -1;			// Prev index
		bool issue = false;	// Did we detected an issue ?

		if(!gpxpos){
			puts(" ----------------------------------------------------------------------");
			printf("| GPX : %s -> %s\n", Gpx->getMin().strLocalTime().c_str(), Gpx->getMax().strLocalTime().c_str());
			printf("| Distance traveled : %.0f m\n", Gpx->getLast().getCumulativeDistance());
			puts(" --------------------------------------------------------------------------------------------------------");
			puts("|    video     |         Beginning         |            End            |            Timestamps           | Status");
			puts("|              | Index | Estrgmt  | CumDis | Index | Estrgmt  | CumDis |       GPX      |       GoPro    |");
			puts(" --------------------------------------------------------------------------------------------------------");
			for(auto &v : videos){
				printf("| %12s | %05d | %8.1f | %6.0f | %05d | %8.1f | %6.0f | %s -> %s | %s -> %s | ",
					basename(v.c_str()),
					v.beginning.idx, v.beginning.distance, (*Gpx)[v.beginning.idx].getCumulativeDistance(),
					v.end.idx, v.end.distance, (*Gpx)[v.end.idx].getCumulativeDistance(),
					(*Gpx)[v.beginning.idx].strLocalHour(true).c_str(), (*Gpx)[v.end.idx].strLocalHour(true).c_str(),
					v.getFirst().strLocalHour(true).c_str(), v.getLast().strLocalHour(true).c_str()
				);

				if(!v.getMax().gfix){
					puts(" No GPS");
					issue = true;
				} else if(!v.getMin().gfix){
					puts(" Partial GPS");
					issue = true;
				} else if(v.getMin().dop > 500){
					puts(" Bad signal");
					issue = true;
				} else if(v.getMax().dop > 500){
					puts(" Weak signal");
					issue = true;
				} else if(v.end.idx == -1){
					puts(" Not ending");
					issue = true;
				} else if(prev > v.beginning.idx){
					puts(" Overlapping");
					issue = true;
				} else
					puts(" ok");

				prev = v.beginning.idx;
			}
		} else {
			puts(" ----------------------------------------------------------------------");
			printf("| GPX : %s -> %s\n", Gpx->getMin().strLocalTime().c_str(), Gpx->getMax().strLocalTime().c_str());
			printf("| Distance traveled : %.0f m\n", Gpx->getLast().getCumulativeDistance());
			puts(" ------------------------------------------------------------------------------------------------------------");
			puts("|    video     |    Beginning     |       End        |       Lenght         |           Timestamps           | Status");
			puts("|              | Index | Estrgmt  | Index | Estrgmt  |    GPX   |   GoPro   |      GPX      |       GoPro    |");
			puts(" ------------------------------------------------------------------------------------------------------------");
			for(auto &v : videos){
				printf("| %12s | %05d | %8.1f | %05d | %8.1f | %8.1f | %8.1f | %s -> %s | %s -> %s | ",
					basename(v.c_str()),
					v.beginning.idx, v.beginning.distance,
					v.end.idx, v.end.distance,
					(*Gpx)[v.end.idx].getCumulativeDistance() - (*Gpx)[v.beginning.idx].getCumulativeDistance(),
					v.getLast().getCumulativeDistance(),
					(*Gpx)[v.beginning.idx].strLocalHour(true).c_str(), (*Gpx)[v.end.idx].strLocalHour(true).c_str(),
					v.getFirst().strLocalHour(true).c_str(), v.getLast().strLocalHour(true).c_str()
				);

				if(!v.getMax().gfix){
					puts(" No GPS");
					issue = true;
				} else if(!v.getMin().gfix){
					puts(" Partial GPS");
					issue = true;
				} else if(v.getMin().dop > 500){
					puts(" Bad signal");
					issue = true;
				} else if(v.getMax().dop > 500){
					puts(" Weak signal");
					issue = true;
				} else if(v.end.idx == -1){
					puts(" Not ending");
					issue = true;
				} else if(prev > v.beginning.idx){
					puts(" Overlapping");
					issue = true;
				} else
					puts(" ok");

				prev = v.beginning.idx;
			}
		}

		if(issue){
			fputs("*W* At least Problem has been detected\n", stderr);
			if(!force){
				fputs("*F* Use '-F' flag to accept it\n", stderr);
				exit(EXIT_FAILURE);
			}
		}

			/* **
			 * Generate the story
			 * **
			 *
			 * As we're relying on previous processing, we don't have to check
			 * boundary
			 */
		fputs("GPMFStory 1.0\n", story);	// Header to identify a story
		fprintf(story, "#GPX starting time : %s\n", Gpx->getMin().strLocalTime().c_str());

		fputs("#Video'name, starting Index, ending Index\n", story);
		for(auto &v : videos)
			fprintf(story, "%s, %5d, %5d\n", basename(v.c_str()), v.beginning.idx, v.end.idx);

		fputs("#GPX data\n*GPX\n", story);
		fputs("#latitude, longitude, altitude, sample_time, cumulative_distance\n", story);
		for( idx = 0; idx < (int)Gpx->getSampleCount(); idx++ ){
			auto &gpx = (*Gpx)[idx];
			fprintf(story, "%f, %f, %f, %lu, %f\n",
				gpx.getLatitude(), gpx.getLongitude(),
				gpx.getAltitude(), gpx.getSampleTime(),
				gpx.getCumulativeDistance()
			);
		}
			

		fclose(story);
	} else {
			/* In this mode, the journey is rebuilt from the succession of videos
			 * (with some potential gaps)
			 */

		if(verbose)
			puts("*I* Videos only mode");

		if(!(story = fopen("story", "w"))){
			perror("story");
			exit(EXIT_FAILURE);		
		}

		fputs("GPMFStory 2.0\n", story);	// Header to identify a story

		bool first = true;
		double last_cumdistance = 0.0;
		GPMFdata *last = NULL;
		for(; optind < argc; optind++){

			GPVideo video(argv[optind], 1, last_cumdistance);	// Read video's data

			if(first){
				first = false;
				auto p = video.getFirst();
				fputs("# starting time :", story);
				printtm(p.getLocalTime(), story);
				fputs("\n", story);
				//				fprintf(story, "*start %lu\n", p.getSampleTime());
				fputs("#latitude, longitude, altitude, sample_time, cumulative_distance\n", story);
			} else {
				double diff = video.getFirst().diffTimeF(last->getSampleTime());
				if(verbose)
					printf("*I* gap : %s min, estrangement : %.2lf m\n",
						video.getFirst().diffTime(last->getSampleTime()).c_str(),
						video.getFirst().Estrangement(*last)
					);
				if(diff < 0){
					printf("*%c* Video returning backward by %s minutes\n",
						force ? 'E':'F',
						video.getFirst().diffTime(last->getSampleTime()).c_str()
					);
					if(!force)
						exit(EXIT_FAILURE);
				}
			}

			for(auto s: video.getSamples()){
				fprintf(story, "%f, %f, %f, %lu, %f\n",
					s.getLatitude(), s.getLongitude(),
					s.getAltitude(), s.getSampleTime(),
					s.getCumulativeDistance()
				);

				last_cumdistance = s.getCumulativeDistance();
				last = &s;
			}

			fprintf(story, "*Video\t%s\n", argv[optind]);

			if(debug)
				video.Dump();
		}

		fclose(story);
	}
}

